README.md
# schedtoold
## original schedtoold, with archlinux patches and systemd unit file.
credit to [original source](http://web.archive.org/web/20150803004758/http://www.darav.de/schedtoold.html), and [archlinux AUR package](https://aur.archlinux.org/packages/schedtoold/) for patches and systemd unit file.


ChangeLog
Version 0.3
- renamed to schedtoold (instead of scheD) for gentoo packaging conformance
- renamed sched.cfg to schedtoold.conf, it's more "standard" conform 
  (Andreas Mohr)
- new features:
  - "-p pid_file" option
  - "-u sec" option to set update interval
  - possibility to define rules per user or group in schedtoold.conf
    look at doc/README and doc/schedtoold.conf-example
  - absolute paths for executables are now supported, like /usr/bin/xmms
    AND STRONGLY RECOMMENDED! Look at doc/README for details.
    (Solution isn't perfect yet. See discussion on ck-mailing list.)
  - gentoo and debian init-scripts (thanks Ray for his contribution)
  - running as non-root is now supported, only processes with schedtoold's
    uid will be touched
- fixes:
  - wildcard support
  - changed error messages of schedtool (Javier Kohen)
  - spelling fixes (Andreas Mohr)

Version 0.2
- support for wildcards (name* , but the '*' must be at the end of programm
name)
- improved config file handling (f.e. no limit for count of configured programms)
- more error checking
- verbose improved (when running -v)
- use pid_t instead of int
- updated sched.cfg example
- added schedtool-1.2.4.patch for newer ck-patchset to handle SCHED_ISO right
- last and least: run indent to format source code :)

Version 0.1
- initial release 


utils.c
/*
 * Copyright (C) 2005 Carsten Rietzschel <crPLESAE@DOdaravNOT.SPAMde>
 *
 * This file is part of the schedtoold, which is distributed under
 * the  terms  of the  GNU General Public Licence 2.  Please see the
 * COPYING file for details.
 */

/* globals */
#include <stdio.h>
#include <stdlib.h>
#include <syslog.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <stdarg.h>

/* local */
#include "utils.h"
#include "config-handler.h"

/* 
 * run as daemon (default: yes) 
 */
int run_daemon = 1;

/*
 * verbose level
 */
int verbose = 0;

/*
 * ---------------------------------------------------------------------------- 
 * helper functions
 * ----------------------------------------------------------------------------  
 */




/*
 * output via syslog (daemon) or printf 
 */
void
msg (char *output, ...)
{
  va_list list;
  char infostr[256];

  va_start (list, output);
  vsnprintf (infostr, 256, output, list);
  va_end (list);
  if (run_daemon)
    syslog (LOG_DAEMON | LOG_INFO, "%s", infostr);
  else
    printf ("%s\n", infostr);
}



/*
 * termination handler (called when exiting)
 */
void
terminationHandler (int signum)
{
  /* free config entries */
  configFreeEntries ();

  /* bye */
  msg ("Terminated. Bye bye.");
  exit (EXIT_SUCCESS);
}


/*
 * write pidfile
 */
void
writePidFile (char *filename)
{
  pid_t pid = getpid ();
  int fd = open (filename, O_TRUNC | O_CREAT | O_WRONLY,S_IRUSR|S_IWUSR);
  char buf[6];

  snprintf (buf, 6, "%i", pid);
  if (fd >= 0)
  {
    /* write pid into */
    write (fd, &buf, strlen (buf));
    /* chmod it and close */
    fchmod (fd, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
    close (fd);
  }
  else
    msg ("Warning! Writing pid_file failed (%s)!\n", strerror (errno));

}


/*
 * become a daemon via fork
 */
void
daemonize (void)
{
  pid_t pid;

  pid = fork ();
  switch (pid)
  {
  case -1:
    msg ("Error! fork sched failed, exiting.");
    exit (EXIT_FAILURE);
  case 0:
    setsid ();
    chdir ("/");
    umask (0);
    break;
  default:
    exit (EXIT_SUCCESS);
  }
}


pid-handler.c
/*
 * Copyright (C) 2005 Carsten Rietzschel <crPLESAE@DOdaravNOT.SPAMde>
 *
 * This file is part of the schedtoold, which is distributed under
 * the  terms  of the  GNU General Public Licence 2.  Please see the
 * COPYING file for details.
 */


/* globals.h */
#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <dirent.h>

/* local */
#include "pid-handler.h"
#include "utils.h"

/*
 * list of pids, pid=PID_FREE means empty slot, 
 * pid=PID_EOL is last element (to stop search as early as possible)
 */
list_entry_t pid_list[PID_MAX];

/* current age to kill closed processes from list */
static int age = 0;


/*
 * dump pid list
 */
void
pidListDump (void)
{
  int i;

  printf ("dump of pid list:\n");
  for (i = 0; i < PID_MAX - 1; i++)
  {
    if (pid_list[i].pid == PID_FREE)
      printf ("\t%3d: free\n", i);
    else if (pid_list[i].pid == PID_EOL)
    {
      printf ("\t%3d: --end of list--\n", i);
      return;
    }
    else
      printf ("\t%3d: PID=%5d Status=%i\n", i, pid_list[i].pid,
	      pid_list[i].status);
  }
}

/*
 * add pid to list
 * return index in pid_list
 */
static int
pidListAdd (int pid)
{
  int i;
  int free_entry = -1;

  for (i = 0; i < (PID_MAX - 1); i++)
  {
    if (pid_list[i].pid == pid)
    {
      /* found */
      pid_list[i].age = age;
      return i;
    }
    else if (pid_list[i].pid == PID_FREE)
    {
      /* remember first free entry */
      if (free_entry == -1)
	free_entry = i;
    }
    else if (pid_list[i].pid == PID_EOL)
    {
      /* reached end and not already found - so we have to add it */
      if (free_entry == -1)
	free_entry = i;

      pid_list[free_entry].pid = pid;
      /* mark for update */
      pid_list[free_entry].status = PID_STATUS_UNSET;
      pid_list[free_entry].age = age;
      return free_entry;
    }
  }
  /* error: too many entries in this list */
  msg ("[pids]   too many entries in in pid-list!");
  return -1;
}


/*
 * remove dead processes from list
 */
static void
pidListRemoveDead (void)
{
  int i;

  for (i = 0; i < (PID_MAX - 1); i++)
  {
    if (pid_list[i].pid == PID_EOL)
    {
      /* increase age for next loop */
      age++;
      return;
    }
    else if ((pid_list[i].age != age) && (pid_list[i].pid != PID_FREE))
    {
      /* it's removed */
      pid_list[i].pid = PID_FREE;
    }

  }
  age++;
}


/*
 * check status of pid 
 * return: PID_SET, PID_NOT_SED or PID_RETRY
 */
int
pidGetStatus (int index)
{
  return pid_list[index].status;
}


/*
 * set status of pid
 */
void
pidSetStatus (int index, int pid_value)
{
  pid_list[index].status = pid_value;
}


/*
 * simple filter out all direntries not beginning with 0-9
 */
static int
pidFilter (const struct dirent *entry)
{
  if ((entry->d_name[0] >= '0') && (entry->d_name[0] <= '9'))
    return 1;
  else
    return 0;
}


/*
 * read all pids from /proc
 * update in-memory list
 * attempt schedtool action
 */
int
pidUpdateViaProc ()
{
  struct dirent **namelist;
  int n;

  n = scandir ("/proc", &namelist, pidFilter, 0);
  if (n < 0)
  {
    perror ("scandir");
    return -1;
  }
  else
  {
    while (n--)
    {
      int pid;

      /* add to pid list */
      pid = atoi (namelist[n]->d_name);
      if (pid > PID_IGNORE)
	/* add to list */
	pidListAdd (pid);

      /* free element */
      free (namelist[n]);
    }
    /* free list */
    free (namelist);
  }

  /* remove dead pids from pid_list */
  pidListRemoveDead ();

  return 0;
}


pid-handler.h
/*
 * Copyright (C) 2005 Carsten Rietzschel <crPLESAE@DOdaravNOT.SPAMde>
 *
 * This file is part of the schedtoold, which is distributed under
 * the  terms  of the  GNU General Public Licence 2.  Please see the
 * COPYING file for details.
 */


#ifndef PID_HANDLER_H
#define PID_HANDLER_H

/* various definitions */
#define PID_MAX  	32768
#define PID_FREE 	-1
#define PID_EOL 	0

#define PID_STATUS_UNSET	0	//not set
#define PID_STATUS_SET		1	//priority for this pid as already set
#define PID_STATUS_RETRY	2	//1st run of schedtool failed - retry

#define PID_IGNORE      1000	//ignore pids smaller PID_IGNORE


typedef struct list_entry
{
  pid_t pid;			// pid
  int status;			// priority already set or no config option to set
  int age;			// each loop increase age to detect closed processes
} list_entry_t;

/* in-memory list of pids in system */
extern list_entry_t pid_list[PID_MAX];


/*
 * dump pid list
 */
void pidListDump (void);

/*
 * check status of pid 
 * return: PID_SET, PID_NOT_SED or PID_RETRY
 */
//inline int pidGetStatus (int index);
int pidGetStatus(int index);

/*
 * set status of pid
 */
//inline void pidSetStatus (int index, int pid_value);
void pidSetStatus(int index, int pid_value);

/*
 * read all pids from /proc
 * update in-memory list
 * attempt schedtool action
 */
//inline int pidUpdateViaProc (void);
int pidUpdateViaProc(void);


#endif


config-handler.c
/*
 * Copyright (C) 2005 Carsten Rietzschel <crPLESAE@DOdaravNOT.SPAMde>
 *
 * This file is part of the schedtoold, which is distributed under
 * the  terms  of the  GNU General Public Licence 2.  Please see the
 * COPYING file for details.
 */

/* globals */
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <syslog.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>

/* local */
#include "config-handler.h"
#include "utils.h"

static config_entry_t config_first_entry = { NULL, NULL, 0, 0, 0, 0, NULL };
static config_entry_t *config_last_entry = &config_first_entry;


/*
 * read from filename to get userIDs or groupIDs from given name
 *
 * returns ID ;  on error: < 0
 * 
 * NOTE: we could read file only once and cached all data, but as it's done only 
 * at startup of schedtoold I'll keep it the simple way :)
 */
static int
configReadID (char *filename, char *name)
{
  FILE *file = fopen (filename, "r");
  char line[256];
  int id = -1;
  char read[256];

  if (file != NULL)
  {
    /* read until we found username */
    while (fgets (line, 255, file) != NULL)
    {
      /* 
       * scan for username:x:uid or groupname:x:gid 
       */
      sscanf (line, "%[a-zA-z0-9]:x:%i:\n", read, &id);
      if (!strncmp (read, name, 32))
      {
	/* found */
	fclose (file);
	return id;
      }
    }

    /* close file */
    fclose (file);
  }
  /* open failed or not found */
  return ID_NOT_FOUND;
}



/*
 * add entry to in-memory config list
 */
static int
configAddEntry (int uid, int gid, char *name, char *options, int wildcard,
		int absolute)
{
  config_entry_t *entry = malloc (sizeof (config_entry_t));
  if (entry == NULL)
    return -1;

  /* verbose */
  if (verbose)
    printf ("Added: %s %s%s (uid:%i,gid:%i)\n",
	    name, options, wildcard ? " (wildcard)" : "",
	    uid, gid);

  /* add strings */
  config_last_entry->next = entry;
  entry->name = name;
  entry->options = options;
  entry->wildcard = wildcard;
  entry->absolute = absolute;
  entry->gid = gid;
  entry->uid = uid;
  entry->next = NULL;

  /* remember last entry */
  config_last_entry = entry;

  /* done */
  return 0;
}


/*
 * free the whole in-memory list of config entries
 */
void
configFreeEntries ()
{
  config_entry_t *entry = config_first_entry.next;
  config_entry_t *next_entry;
  while (entry)
  {
    next_entry = entry->next;
    /* free strings */
    free (entry->name);
    free (entry->options);

    /* free entrie itself */
    free (entry);

    entry = next_entry;
  }
  config_first_entry.next = NULL;
}



/* 
 * reads configfile 
 */
int
configLoadFile (char *filename)
{
  FILE *file;
  char line[256];
  char scan_name[128];
  int linecount = 0;

  char *name;
  char *options;
  int options_index;		/* index of options in read line */
  int wildcard;
  int absolute;
  int uid;			/* user id */
  int gid;			/* group id */

  /* open */
  file = fopen (filename, "r");
  if (file == NULL)
  {
    printf ("Error! Reading configfile %s failed!\n", filename);
    return -1;
  }

  /* now read */
  while (fgets (line, 256, file) != NULL)
  {
    linecount++;

    /* ignore comments */
    if (line[0] == '#')
      continue;

    /* kill newline at end of line */
    line[strlen (line) - 1] = '\0';

    /* parse for name */
    scan_name[0] = '\0';
    sscanf (line, "%127s", scan_name);
    scan_name[127] = '\0';

    /* begin of options relavtive to line just read */
    options_index = strlen (scan_name);

    /* check if user or group is set */
    if (scan_name[options_index - 1] == ':')
    {
      /* found a user- or groupname */

      /* kill ':' */
      scan_name[options_index - 1] = '\0';

      /* group or user? */
      if (scan_name[0] == '%')
      {
	/* get group id */
	if (!strcmp (&scan_name[1], "ALL"))
	  gid = GID_ALL;
	else
	  gid = configReadID ("/etc/group", &scan_name[1]);

	/* valid for all users */
	uid = UID_ALL;
      }
      else
      {
	/* get user id */
	if (!strcmp (scan_name, "ALL"))
	  uid = UID_ALL;
	else
	  uid = configReadID ("/etc/passwd", scan_name);

	/* valid for all groups */
	gid = GID_ALL;
      }

      if ((uid != ID_NOT_FOUND) && (gid != ID_NOT_FOUND))
      {
	/* found - update scan_name */
	sscanf (&line[strlen (scan_name) + 1], "%127s", scan_name);
      }
      else
      {
	/* not found - invalid scanname to catch error later! */
	scan_name[0] = '\0';
	printf
	  ("Warning! Line %i in config file contains invalid user or group. Ignored.\n",
	   linecount);
      }

      /* update options postion */
      options_index += strlen (scan_name);

    }
    else
    {
      /* no user or group set: relevant for all users of all groups! */
      uid = UID_ALL;
      gid = GID_ALL;
    }

    /* create strings to be stored in in-memory list */
    name = strdup (scan_name);
    options = strdup (&line[options_index + 1]);

    /* to avoid stupid errors */
    if ((name == NULL) || (options == NULL))
    {
      printf ("Error! Malloc failed. Abort reading config file.\n");
      if (name)
	free (name);
      if (options)
	free (options);
      break;
    }

    /* valid? (name and at least one option) */
    if ((strlen (options) > 0) && (strlen (name) > 0))
    {
      /* 
       * check for wildcard 
       * XXX: the only valid wildcard is programname* (the * at the end!) 
       */
      if (name[strlen (name) - 1] == '*')
	wildcard = 1;
      else
	wildcard = 0;

      /* 
       * check for absolute paths to executable 
       */
      if (name[0] == '/')
	absolute = 1;
      else
	absolute = 0;

      /* add to in-memory list */
      if (configAddEntry (uid, gid, name, options, wildcard, absolute))
      {
	printf ("Error! Malloc failed. Abort reading config file.\n");
	if (name)
	  free (name);
	if (options)
	  free (options);
	break;
      }

    }
    else
    {
      /* invalid line */
      if (strlen (name) > 0)
	printf
	  ("Warning! Line %i in config file has no schedtool options.\n",
	   linecount);

      /* free strings */
      if (name)
	free (name);
      if (options)
	free (options);
      name = options = NULL;

    }


  }
  /* close */
  fclose (file);
  /* done */
  return 0;
}


/*
 * compare given statusname with names from config 
 * to handle absolute paths we need the path to the executable (exename)
 *
 * returns index in config_list if found
 * on error returns value <0
 */
config_entry_t *
configSearch (char *statusname, char *exename, uid_t uid, gid_t gid)
{
  config_entry_t *entry = config_first_entry.next;
  char *comparename = statusname;

  while (entry)
  {
    config_entry_t *found = NULL;


    if ((entry->absolute) && (strlen (exename) > 0))
      /* 
       * for config with full path and valid exename (symlink readable) 
       * we take take the exename 
       */
      comparename = exename;
    else
      /* 
       * compare the whole statusname if no path given in config
       * or no valid exename was found
       */
      comparename = statusname;

    /* compare */
    if (entry->wildcard)
    {
      /* for wildcard we compare all characters except the final '*'  */
      if (!strncmp (entry->name, comparename, strlen (entry->name) - 1))
	found = entry;
    }
    else
    {
      /* full string */
      if (!strcmp (entry->name, comparename))
	found = entry;
    }

    /* found? then compare gid and uid */
    if (found)
    {
      if (((entry->gid == GID_ALL) || (entry->gid == gid))
	  && ((entry->uid == UID_ALL) || (entry->uid == uid)))
	return found;

      if (verbose)
	msg ("Warning! uid/gid mismatch for %s (exe: %s). Ignored.",
	     statusname, exename);
    }

    /* next one */
    entry = entry->next;
  }

  /* not found */
  return NULL;
}


config-handler.h
/*
 * Copyright (C) 2005 Carsten Rietzschel <crPLESAE@DOdaravNOT.SPAMde>
 *
 * This file is part of the schedtoold, which is distributed under
 * the  terms  of the  GNU General Public Licence 2.  Please see the
 * COPYING file for details.
 */

#ifndef LOADCONFIG_H
#define LOADCONFIG_H


#define UID_ALL	-1
#define GID_ALL	-1
#define ID_NOT_FOUND -2

/* in-memory list of out config */
typedef struct config_entry
{
  char *name;			//processname
  char *options;		//schedtool command line options
  int wildcard;			//flag set if config entry is a wildcard
  int absolute;			//flag is set if an absolute path to the executable is given
  uid_t uid;			//relevant user id
  gid_t gid;			//relevant group id
  void *next;			//next config entry
} config_entry_t;


/* 
 * reads configfile 
 */
int configLoadFile (char *filename);

/* 
 * search name in list 
 */
config_entry_t *configSearch (char *statusname, char *exename, uid_t uid,
			      gid_t gid);

/*
 * free config entries 
 */
void configFreeEntries (void);

#endif


main.c
/*
 * Copyright (C) 2005 Carsten Rietzschel <crPLESAE@DOdaravNOT.SPAMde>
 *
 * This file is part of the schedtoold, which is distributed under
 * the  terms  of the  GNU General Public Licence 2.  Please see the
 * COPYING file for details.
 */

/* globals */
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <syslog.h>
#include <fcntl.h>
#include <unistd.h>
#include <string.h>

/* local */
#include "utils.h"
#include "pid-handler.h"
#include "config-handler.h"

#define SCHED_VERSION "0.3"


//warning XXX - Make us non-global - please :)
static char exename[1024];	/* name with path from /proc/pid/exe */
static char statusname[1024];	/* name from /proc/pid/status */
static uid_t proc_uid;		/* uid of pid */
static gid_t proc_gid;		/* gid of pid */


/*
 * prints usage info 
 */
void
usage (void)
{
  printf
    ("schedtoold %s: daemon to automagically update priority/nice-level of running programs.\n",
     SCHED_VERSION);
  printf (" [-h]\t\tshow this help\n"
	  " [-v]\t\tbe more verbose\n"
	  " [-n]\t\tstay in foreground (no daemon)\n"
	  " [-u sec]           update interval in seconds\n"
	  " [-c config_file]   load given config file (default: /etc/schedtoold.conf)\n"
	  " [-p pid_file]      filename to store pid in\n");
  exit (EXIT_SUCCESS);
}


/*
 * get process info from proc
 * return: 0 on sucess
 * error: -1 
 */
static inline int
getProcInfo (pid_t pid)
{
  int fd;
  int exename_len;

  /*
   * We get the name of program is it was started from /proc/pid/status
   * and we also get the exe-symlink which is running.
   * This is necessary for all scripts:
   *   /proc/pid/status shows the name as it was started (e.g. scriptname)
   *   and /proc/pid/exe is symlinked e.g. to bash
   * If we found one script running we're looking into /proc/pid/cmdline
   * and trying to get the absolute path to the 
   *
   * XXX - this is not perfect :(
   * Any ideas how to get the absolute path of the program (script,binary,..)?
   */

  /*
   *  get name from /proc/pid/status 
   */
  snprintf (statusname, 1023, "/proc/%i/status", pid);
  fd = open (statusname, O_RDONLY);
  if (fd > 0)
  {
    char buf[512];
    char *idstart;

    if (read (fd, buf, 511) > 0)
    {
      if (sscanf (buf, "Name:\t%511s\n", statusname) != 1)
      {
	/* error in conversation */
	close (fd);
	return -1;
      }

      /* check UID */
      idstart = strstr (buf, "Uid:");
      if ((idstart == NULL)
	  || (sscanf (idstart, "Uid:\t%i\t", &proc_uid) != 1))
      {
	/* either error in conversation  */
	close (fd);
	return -1;
      }

      /* get GID */
      idstart = strstr (buf, "Gid:");
      if ((idstart == NULL)
	  || (sscanf (idstart, "Gid:\t%i\t", &proc_gid) != 1))
      {
	/* either error in conversation  */
	close (fd);
	return -1;
      }

    }
    close (fd);
  }
  else
    statusname[0] = '\0';	/* make name read invalid */


  /*
   * get name from /proc/pid/exe-symlink and if necessary also get
   * /proc/pid/cmdline
   */
  snprintf (exename, 1023, "/proc/%i/exe", pid);
  char *buf;
  buf = malloc(1023);
  exename_len = readlink(exename, buf, 1023);
  //if ((exename_len = readlink (exename, exename, 1023)) > 0)
  if(exename_len > 0)
  {
    char *binaryname;		/* the binary without path from exename */
    char cmdline[1024];		/* proc/pid/cmdline */
    int cmdline_len;		/* bytes read */

    /* readlink success and append final NULL to string */
    exename[exename_len] = '\0';
    binaryname = strrchr (exename, '/');
    binaryname++;
    /* compare exe- vs. statusname to detect scripts */
    if (strcmp (binaryname, statusname))
    {
      snprintf (cmdline, 1023, "/proc/%i/cmdline", pid);
      fd = open (cmdline, O_RDONLY);
      if (fd > 0)
      {
	if ((cmdline_len = read (fd, &cmdline, 1023)) > 0)
	{
	  /* 
	   * for all params we check if we found status name - hoping it has an 
	   * absolute path given 
	   */
	  char *param = cmdline;

	  while (param != NULL)
	  {
	    if ((cmdline_len -= strlen (param) + 1) <= 0)
	      break;

	    /* is this the param containing the status name? */
	    binaryname = strrchr (param, '/');
	    if (binaryname != NULL)
	    {
	      binaryname++;
	      if (!strcmp (binaryname, statusname))
	      {
		strncpy (exename, param, 1024);
		break;
	      }
	    }

	    /* get next param */
	    param = (char *) memchr (param, '\0', cmdline_len);
	    if (param != NULL)
	      param++;		/* the charater after \0 is relevant */

	  }
	}
	close (fd);
      }
    }
  }
  else
    exename[0] = '\0';		/* make name read invalid */

  /* sucess */
  return 0;
}


/*
 * just the main w/ work loop
 */
int
main (int argc, char *argv[])
{
  int c;
  char *config_file = NULL;
  char *pid_file = NULL;
  uid_t current_uid;		/* uid of schedtoold itself */
  unsigned int update_sec = 2;	/* time between to updates in seconds */
  char *buf = exename;		/* don't waste memory but make code a bit more readable */


  /* check if we're running as root, if not warn */
  if ((current_uid = getuid ()) != 0)
  {
    printf
      ("Warning! Not running with UID=0 (root)!\n         Setting priorities normally requires root privileges.\n");
    printf ("Only programs with UID=%i will be touched by sched.\n\n",
	    current_uid);
  }


  /* process cmdline arguments */
  while ((c = getopt (argc, argv, "vhnc:p:u:")) != -1)
  {
    switch (c)
    {
    case 'n':
      run_daemon = 0;
      break;
    case 'v':
      verbose++;
      break;
    case 'h':
      usage ();
      break;
    case 'c':
      config_file = optarg;
      break;
    case 'p':
      pid_file = optarg;
      break;
    case 'u':
      update_sec = atoi (optarg);
      if (update_sec < 1)
      {
	printf ("Error! Update interval must be at least one second!\n");
	exit (EXIT_FAILURE);
      }
      break;
    case '?':
      usage ();
      break;
    }
  }


  /* load config */
  if (config_file == NULL)
  {
    /* no config give via cmdline */
    if (current_uid == 0)
      config_file = "/etc/schedtoold.conf";
    else
    {
      /* non-root user get it from home */
      snprintf (buf, 128, "%s/.schedtoold.conf", getenv ("HOME"));
      config_file = buf;
    }
  }

  if (configLoadFile (config_file))
    exit (EXIT_FAILURE);

  /* now we daemonize */
  if (run_daemon)
    daemonize ();

  /* write pid file */
  if (pid_file)
    writePidFile (pid_file);

  /* Invoke terminationHandler when one of these signals happens */
  signal (SIGTERM, terminationHandler);
  signal (SIGINT, terminationHandler);
  signal (SIGHUP, terminationHandler);
  signal (SIGQUIT, terminationHandler);

  /* verbose into syslog */
  msg ("Ready: version %s", SCHED_VERSION);

  /* main loop */
  while (1)
  {
    int i;

    /* update list */
    pidUpdateViaProc ();

    /* check which pids should be updated */
    for (i = 0; i < PID_MAX - 1; i++)
    {
      /* end of list? */
      if (pid_list[i].pid == PID_EOL)
	break;

      /* needs to be updated? */
      if ((pidGetStatus (i) != PID_STATUS_SET)
	  && (pid_list[i].pid != PID_FREE))
      {
	config_entry_t *config_entry;	/* config entry */

	/* 
	 * get info from /proc for pid
	 */
	if (getProcInfo (pid_list[i].pid) != 0)
	{
	  /* couldn't get /proc-infos  */
	  config_entry = NULL;	/* next one */
	  if (verbose)
	    msg ("Error: Couldn't get info from /proc/%i\n", pid_list[i].pid);
	}
	else
	{
	  if ((current_uid == 0) || (current_uid == proc_uid))
	    /* got info - search in out config */
	    config_entry =
	      configSearch (statusname, exename, proc_uid, proc_gid);
	  else
	    /* not running as root and uid doesn't match */
	    config_entry = NULL;
	}

	if (config_entry != NULL)
	{
	  /* we found a valid one to update its priority */

	  /* verbose */
	  if (verbose)
	    msg
	      ("Updating: %s%s (exe:%s, pid:%i, options:%s)",
	       statusname,
	       pidGetStatus (i) == PID_STATUS_RETRY ? " (retry)" : "", exename,
	       pid_list[i].pid, config_entry->options);

	  /* run schedtool */
	  snprintf (buf, 1023, "schedtool %s %i",
		    config_entry->options, pid_list[i].pid);
	  if (system (buf))
	  {
	    /* show error message only at 2nd try - cause program could be already exited when trying 1st time */
	    if (pidGetStatus (i) == PID_STATUS_RETRY)
	    {
	      msg
		("Error! Running \"schedtool %s\" on %s (pid: %i) failed.",
		 config_entry->options, statusname, pid_list[i].pid);
	      /* don't try again */
	      pidSetStatus (i, PID_STATUS_SET);
	    }
	    else
	      /* retry next time */
	      pidSetStatus (i, PID_STATUS_RETRY);
	  }
	  else
	    /* success: mark set */
	    pidSetStatus (i, PID_STATUS_SET);

	}
	else
	  /* not found - but mark as set to avoid re-check */
	  pidSetStatus (i, PID_STATUS_SET);
      }
    }

    /* wait until a process is created/closed */
    sleep (update_sec);
    /* ?? XXX - is there a better way than polling ?? */
  }

  /* bye */
  msg ("Have a nice day.");

  /* free memory */
  configFreeEntries ();

  return 0;
};
